### PR TITLE
Fix FEM dynamic cache identities

### LINKED ISSUE #1866: `warp.fem.ImplicitField` returns silently wrong values when two fields share a domain
## Summary

`fem.ImplicitField.name` does not encode the identity of the field's `func`, but the
generated device functions are cached on that name. Two `ImplicitField`s that share
`(domain, degree, values-signature)` but wrap **different** `wp.func`s therefore share the
**first** one's compiled evaluator. The second field silently evaluates the first field's
function.

There is no error, no warning, and the result is plausible — it is simply the wrong
coefficient. The condition persists for the lifetime of the process: a *newly constructed*
field with the second function is also wrong.

## Reproducer

```python
import warp as wp
import warp.fem as fem


@wp.func
def f_one(x: wp.vec2d):
    return wp.float64(1.0)


@wp.func
def f_two(x: wp.vec2d):
    return wp.float64(2.0)


@fem.integrand
def integrate_coefficient(s: fem.Sample, c: fem.Field):
    return fem.inner(c, s)


geo = fem.Grid2D(res=wp.vec2i(2, 2), scalar_type=wp.float64)
cells = fem.Cells(geo)
quadrature = fem.RegularQuadrature(domain=cells, order=2)

field_one = fem.ImplicitField(domain=cells, func=f_one, degree=0)
field_two = fem.ImplicitField(domain=cells, func=f_two, degree=0)

print("field_one.name:", field_one.name)
print("field_two.name:", field_two.name)
print("names collide:", field_one.name == field_two.name)

for label, field, expected in (("f_one", field_one, 1.0), ("f_two", field_two, 2.0)):
    got = fem.integrate(integrate_coefficient, quadrature=quadrature,
                        fields={"c": field}, output_dtype=wp.float64)
    print(f"integral over unit square of {label}: got {got}, expected {expected}")

fresh = fem.ImplicitField(domain=cells, func=f_two, degree=0)
got = fem.integrate(integrate_coefficient, quadrature=quadrature,
                    fields={"c": fresh}, output_dtype=wp.float64)
print(f"freshly constructed field over f_two: got {got}, expected 2.0")
```

## Actual output

```
field_one.name: Implicit_Grid2D_f64_Cells_0_Args_2babaa9e
field_two.name: Implicit_Grid2D_f64_Cells_0_Args_2babaa9e
names collide: True
integral over unit square of f_one: got 1.0, expected 1.0
integral over unit square of f_two: got 1.0, expected 2.0      <-- WRONG
freshly constructed field over f_two: got 1.0, expected 2.0    <-- WRONG
```

## Expected output

`f_two` should integrate to `2.0` over the unit square. Swapping the construction order of
`field_one` and `field_two` swaps which function "wins", confirming it is first-construction
order and not a property of either function.

## Root cause

`warp/_src/fem/field/field.py:385-388`:

```python
@property
def name(self) -> str:
    """Unique name encoding the domain and argument structure."""
    return f"Implicit_{self.domain.name}_{self.degree}_{self.EvalArg.key}"
```

The docstring states the name encodes "the domain and argument structure" — and that is
exactly the problem: it omits `self._func` (and `_grad_func` / `_div_func`). Two fields with
the same domain, degree and `EvalArg` signature produce an identical name.

`warp/_src/fem/cache.py:375-397` (`setup_dynamic_attributes`) then keys the generated
attributes on that name:

```python
if key is None:
    key = obj.name
...
key = (key, frozenset(constructors.keys()))
...
attrs = cls._cached_dynamic_attrs.get(key)
if attrs is None:
    ...  # build eval_inner / eval_grad_inner / ... from THIS obj
    cls._cached_dynamic_attrs[key] = attrs
else:
    for k, v in attrs.items():
        setattr(obj, k, v)   # <-- second field adopts the first field's evaluators
```

Because `_cached_dynamic_attrs` lives on the **class**, the stale mapping outlives any
individual instance, which is why a freshly constructed field is also affected.

## Additional consequence: `grad_func` leaks both ways (both directions measured)

`_dynamic_attribute_constructors` includes `eval_grad_inner` (built from `self._grad_func`),
so the gradient evaluator is part of the per-name cached attribute set. Consequently,
construction order decides which field's gradient EVERY same-named field gets:

* **with-`grad_func` field built first, grad-less sibling second:** `fem.grad` on the
  grad-less field does not raise — it silently returns the *sibling's* gradient. Measured:
  a field over `7*y` (true ∂/∂x = 0, and autodiff unsupported so the correct behaviour is an
  error) returned `3.0`, the gradient of its sibling `3*x`.
* **grad-less field built first, with-`grad_func` field second:** the field that *supplied*
  a gradient loses it — `fem.grad` on it fails with
  `TypeError: Operator 'grad' is not defined for Field Implicit_...`, despite `grad_func`
  having been passed.

So omitting `grad_func` is not a guard, and supplying it is not a guarantee.

## Impact

Any code building more than one implicit coefficient field on the same domain is exposed.
This is a natural pattern — separate implicit fields for distinct coefficients of one weak
form — and the failure mode is a plausible wrong answer rather than a crash, so it can
survive a green test suite. We hit it while using `ImplicitField` to supply two different
closed-form coefficients to one finite-element form.

## Suggested fix

Include the wrapped functions in the name, e.g.:

```python
@property
def name(self) -> str:
    grad_key = "" if self._grad_func is None else self._grad_func.key
    div_key = "" if self._div_func is None else self._div_func.key
    return (f"Implicit_{self.domain.name}_{self.degree}_{self.EvalArg.key}"
            f"_{self._func.key}_{grad_key}_{div_key}")
```

With this override applied in a subclass, the reproducer above yields `1.0` and `2.0` as
expected, on both `Cells` and `Sides` domains. Note that a subclass also has to reimplement
`trace()`, because `ImplicitField.trace()` (`field.py:447-459`) hardcodes the base class —
so a user-side workaround is possible but awkward, and the fix belongs upstream.

It may also be worth auditing other `FieldLike` implementations for the same pattern, since
the vulnerability is "`name` used as a cache key while omitting a member that changes
generated code" rather than anything specific to `ImplicitField`.

For completeness: the class docstring documents no uniqueness requirement on `func` per
(domain, degree, signature), so this does not appear to be intended-but-undocumented
behaviour; and swapping the construction order of the two fields in the reproducer swaps
which function "wins" (verified), confirming pure first-construction-wins caching.

### System Information

**Component:** `warp.fem`
**Version:** warp-lang 1.16.0 (PyPI wheel); **still present on `main` as of 2026-08-28**
(same code: `warp/_src/fem/field/field.py:386-388`, `warp/_src/fem/cache.py:376-381`).
No existing issue mentions `ImplicitField` (tracker searched 2026-08-28).
**Platform:** Linux aarch64, CPU device, Python 3.14.4 (device-independent — it is a
host-side caching issue)
**Severity:** silent incorrect numerical results from public API, no warning or error

### PR DESCRIPTION
**Pull request by shi-eric**
## Description

FEM caches dynamically generated evaluators under object-derived names. `ImplicitField`, `Hexmesh`, and `PointBasisSpace` omitted function identities or evaluation options from those names, allowing distinct objects to reuse incompatible generated code. Depending on construction order, this could produce incorrect values, expose unsupported gradients or divergences, or suppress supported ones.

Closes #1866.

## Changes

- Add an identifier-safe cache key derived from a Warp function’s global identity.
- Include value, gradient, and divergence functions in `ImplicitField` cache identities.
- Include the cell evaluation mode in `Hexmesh` cache identities.
- Include kernel and optional gradient functions in `PointBasisSpace` cache identities.
- Add CPU and CUDA regressions covering observable interpolation behavior, both optional-function construction orders, same-named functions from different modules, and distinct Hexmesh evaluation modes.
- Correct the `Hexmesh` constructor parameter name in its docstring.
- Add a changelog fragment.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/warp/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] I added a changelog fragment if this change affects users.

## Validation summary

- Temporarily restored the old `ImplicitField` cache identity and confirmed the four construction-order regressions detect the bug through user-facing behavior: unsupported gradient or divergence operations are wrongly accepted, while supported operations are wrongly rejected.
- Restored the fix and ran all 22 `TestFemField` tests successfully on CPU and CUDA.
- Verified the new Hexmesh and point-basis cache regressions on CPU and CUDA in targeted FEM geometry and quadrature runs.
- Ran the changed-file pre-commit hooks successfully.
- No graph-capture tests were added. The full test suite was not run because validation was limited to the affected FEM components.

## Bug fix

```python
import numpy as np

import warp as wp
import warp.fem as fem


@wp.func
def constant_one(x: wp.vec2):
    return 1.0


@wp.func
def constant_two(x: wp.vec2):
    return 2.0


geometry = fem.Grid2D(res=wp.vec2i(2))
domain = fem.Cells(geometry)
destination = fem.make_polynomial_space(geometry).make_field()

fem.interpolate(fem.ImplicitField(domain, func=constant_one), dest=destination)
fem.interpolate(fem.ImplicitField(domain, func=constant_two), dest=destination)

# Before this fix, the second field could reuse the first field's evaluator.
np.testing.assert_array_equal(destination.dof_values.numpy(), np.full(9, 2.0))
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented incompatible evaluator reuse in finite element workflows.
  * Improved cache separation for implicit fields with different evaluation functions or options.
  * Ensured point-basis spaces distinguish evaluators with different gradients, modules, or configurations.
  * Distinguished hexahedral mesh evaluation modes to prevent incorrect results when switching between parallelepiped and general-cell handling.

* **Documentation**
  * Updated hexahedral mesh constructor documentation to reflect the current evaluation-mode option.

### CODE REVIEW

File: warp/_src/fem/geometry/hexmesh.py [hunk 1]
```diff
@@ -181,7 +181,7 @@ def __init__(
             hex_vertex_indices: warp array of shape (num_hexes, 8) containing vertex indices for each hex
                 following standard ordering (bottom face vertices in counter-clockwise order, then similarly for upper face)
             positions: warp array of shape (num_vertices, 3) containing 3d position for each vertex
-            assume_parallelepiped: If true, assume that all cells are parallelepipeds (cheaper position/gradient evaluations)
+            assume_parallelepiped_cells: If true, assume that all cells are parallelepipeds (cheaper position/gradient evaluations)
```
[the remark this answers is not in the document]
Comment (shi-eric) [reply warp/_src/fem/geometry/hexmesh.py:184]: @coderabbitai The constructor-level parameter documentation predates this PR; this change only corrects the parameter name to match the existing signature. Moving the full `Args` section would be a broader documentation refactor unrelated to the cache-identity fix, so I’m leaving it unchanged here.

File: warp/_src/fem/cache.py [hunk 2]
```diff
@@ -24,6 +25,12 @@
 _key_re = re.compile("[^0-9a-zA-Z_]+")
 
 
+def function_key(func: wp.Function) -> str:
+    """Return an identifier-safe key derived from a Warp function's stable content identity."""
+    uid = ModuleHasher((), {}).hash_function(func)[:16].hex()
+    return f"{_key_re.sub('', func.key)}_{uid}"
```
[the remark this answers is not in the document]
Comment (shi-eric) [reply warp/_src/fem/cache.py]: @coderabbitai `ModuleHasher.hash_function()` already returns `h` at `warp/_src/context.py:3004`, including on commit `2f27358bc`. The amended branch also passed 24 targeted CPU/CUDA tests that construct the affected `ImplicitField` and `PointBasisSpace` objects, so `None[:16]` does not occur and no code change is needed.
